flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java

```diff
@@ -205,20 +208,7 @@ public void onGrantLeadership(UUID newLeaderSessionId) {
     public void onRevokeLeadership() {
         synchronized (lock) {
             if (running) {
-                LOG.debug(
-                        "Revoke leadership of {} ({}@{}).",
-                        leaderContender.getDescription(),
-                        confirmedLeaderInformation.getLeaderSessionID(),
-                        confirmedLeaderInformation.getLeaderAddress());
-
-                issuedLeaderSessionID = null;
-                clearConfirmedLeaderInformation();
-
-                leaderContender.revokeLeadership();
-
-                LOG.debug("Clearing the leader information on {}.", leaderElectionDriver);
-                // Clear the old leader information on the external storage
-                leaderElectionDriver.writeLeaderInformation(LeaderInformation.empty());
```

Comment:
Why is it fine to remove this? The test states that this is done because leadership was already lost.
Does that imply that from the perspective of a TM there is always a leader, because this information is never _cleared_ but only over-written?

Reply:
`onRevokeLeadership()` is called _after_ the leadership is lost. All `LeaderElectionDriver` implementations do a check before writing data to the backend which prevents any writes if the leadership is lost:
* [ZooKeeperLeaderElectionDriver:198](https://github.com/apache/flink/blob/d76214de330fea485f2971b51c172420dfdf500b/flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionDriver.java#L198) (but we even ignore empty leader information in [ZooKeeperLeaderElectionDriver:190](https://github.com/apache/flink/blob/d76214de330fea485f2971b51c172420dfdf500b/flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionDriver.java#L190)). Same code is available for the [ZooKeeperMultipleComponentLeaderElectionDriver.java:154](https://github.com/apache/flink/blob/8ddfd590ebba7fc727e79db41b82d3d40a02b56a/flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java#L154]
* [KubernetesLeaderElectionDriver:139](https://github.com/apache/flink/blob/bba7c417217be878fffb12efedeac50dec2a7459/flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionDriver.java#L139)
* [KubernetesMultipleComponentLeaderElectionDriver:169](https://github.com/apache/flink/blob/0290715a57b8d243586ab747b0cd2416c8081012/flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionDriver.java#L169)

One could argue that the leadership could be regained. But in that case, leadership will be renegotiated anyway which will trigger a write operation with the updated leader information soon after.